src/h.js

```diff
@@ -1,3 +1,24 @@
+var nodePool = []
+
+function reuseNode(node, tag, data, children) {
+  node.tag = tag
+  node.data = data
+  node.children = children
+  return node
+}
+
+export function getNode(tag, data, children) {
+  return nodePool.length > 0 
+    ? reuseNode(nodePool.pop(), tag, data, children)
+    : { tag: tag, data: data, children: children }
+}
+
+export function recoverNode(node) {
+  if (typeof node !== "string") {
+    nodePool.push(node)
```

Comment:
For some extra perfs (and bytes?) I would use a "only growing stack" like in the initial gist.
Meaning that I would avoid the couple `push` / `pop` and only use `nodePool[poolHead]`.

Doing so reduce array internal re-allocations to zero.